### PR TITLE
Major Refactoring to Itinerary 

### PR DESCRIPTION
**Pull request by Sukrut1881**
1) Made itinerary storage similar in concept to HC's ProfileCard

2) Added Tests for all Itinerary Commands

3) Fixed issues regarding date's in recommend and Invalid Itineraries (Please see if this is correctly done as I am not sure if the exceptions are thrown in correct places)

4) More issues to be sorted soon - please git pull and test if your features are working (they should be)

### CODE REVIEW

File: src/main/java/duke/logic/commands/NewItineraryCommand.java [hunk 1]
```diff
@@ -50,47 +42,9 @@ public NewItineraryCommand(LocalDateTime start, LocalDateTime end, String hotel,
...
-            throw new ParseException(Messages.ITINERARY_INCORRECT_COMMAND);
-        }
-        itinerary.setTasks(agendaList);
-
-
-        model.saveItinerary(itinerary);
-        model.itineraryListSave(itinerary);
-        return new CommandResultText("New Itinerary Created :" + itinerary.printItinerary());
+        itinerary.getNumberOfDays();
+        model.storeNewItinerary(itinerary, itineraryDetails);
```
Comment (Jefferson111) [inline src/main/java/duke/logic/commands/NewItineraryCommand.java]: For setters, don't use `storeNewItinerary`, use `setNewItinerary`

File: src/main/java/duke/logic/commands/RecommendationsCommand.java [hunk 2]
```diff
@@ -32,20 +24,13 @@ public RecommendationsCommand(String... itineraryDetails) {
...
-        List<Agenda> list = model.getRecommendations(itinerary.getNumberOfDays(), itinerary);
 
-        assert (!list.isEmpty()) : "list should not be null";
+        Recommendation recommendation = model.getRecommendations();
 
-        itinerary.setTasks(list);
+        Itinerary recentItinerary = recommendation.makeItinerary(itineraryDetails);
 
-        String result = itinerary.printItinerary();
+        model.storeRecentItinerary(recentItinerary);
```
Comment (Jefferson111) [inline src/main/java/duke/logic/commands/RecommendationsCommand.java]: Follow setter naming convention

File: src/main/java/duke/model/ModelManager.java [hunk 3]
```diff
@@ -83,66 +85,53 @@ public EventList getSortedList() {
...
-    /**
-     * Save the newly created itinerary to the Itinerary Lists table of contents.
-     *
-     * @param itinerary The itinerary to be stored.
-     */
     @Override
-    public void itineraryListSave(Itinerary itinerary) throws FileNotSavedException {
-        storage.writeItinerarySave(itinerary);
+    public void storeNewItinerary(Itinerary itinerary, String[] itineraryDetails) throws ParseException {
+        storage.storeNewItinerary(itinerary, itineraryDetails);
```
Comment (Jefferson111) [inline src/main/java/duke/model/ModelManager.java]: This one is still bad

File: src/main/java/duke/storage/Storage.java [hunk 4]
```diff
@@ -289,182 +335,110 @@ private void writeRoutes() throws FileNotSavedException {
...
-                agendaList.add(agenda);
-            }
-            scanner.close();
-            itinerary.setTasks(agendaList);
-        } catch (FileNotFoundException | ParseException e) {
-            throw new FileLoadFailException(SAMPLE_RECOMMENDATIONS_FILE_PATH);
-        }
-        return itinerary;
+    public Itinerary getItinerary(String name) {
+        return itineraryTable.get(name);
```
Comment (Jefferson111) [inline src/main/java/duke/storage/Storage.java]: Is there any class using this method?

File: src/main/java/duke/storage/Storage.java [hunk 5]
```diff
@@ -80,8 +82,57 @@ private void read() throws FileLoadFailException {
...
+                    } else {
+                        break;
+                    }
+                }
+                itinerary.setTasks(agendaList);
+                itineraryTable.put(itinerary.getName(), itinerary);
+            }
+            scanner.close();
+        } catch (FileNotFoundException | ParseException e) {
+            itineraryTable = new HashMap<>();
```
Comment (Jefferson111) [inline src/main/java/duke/storage/Storage.java]: This `itineraryTable = new HashMap<>();` could be moved to the constructor

File: src/main/java/duke/model/planning/Recommendation.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+        List<Agenda> agendaList1 = new ArrayList<>(agendaList.subList(0,getNumberOfDays(start, end)));
+
+        assert (!agendaList1.isEmpty()) : "list should not be null";
+        itinerary.setTasks(agendaList1);
+
+        return itinerary;
+    }
+
+    private int getNumberOfDays(LocalDateTime start, LocalDateTime end) throws
```
Comment (Inno97) [inline src/main/java/duke/model/planning/Recommendation.java]: Catch and handle the API exception.
If it occurs, find some way to atleast have something functional.
Like make a new temporary hotelLocation or something. Don't just let the API exception be thrown all the way up.

File: src/main/java/duke/model/ModelManager.java (hunk 3 above)
Comment (Sukrut1881) [reply src/main/java/duke/model/ModelManager.java]: Fixed

File: src/main/java/duke/storage/Storage.java (hunk 4 above)
Comment (Sukrut1881) [reply src/main/java/duke/storage/Storage.java]: Nope softDeleted

File: src/main/java/duke/storage/Storage.java (hunk 5 above)
Comment (Sukrut1881) [reply src/main/java/duke/storage/Storage.java]: Done

File: src/main/java/duke/logic/commands/RecommendationsCommand.java (hunk 2 above)
Comment (Sukrut1881) [reply src/main/java/duke/logic/commands/RecommendationsCommand.java]: Fixed

File: src/main/java/duke/logic/commands/NewItineraryCommand.java (hunk 1 above)
Comment (Sukrut1881) [reply src/main/java/duke/logic/commands/NewItineraryCommand.java]: Done

File: src/main/java/duke/model/planning/Recommendation.java (hunk 6 above)
Comment (Sukrut1881) [reply src/main/java/duke/model/planning/Recommendation.java]: Yes this was one of the issues raised in the PE. I started by doing this in the recommend command. Will do it in the newItinerary soon

**Pull request merged**